File: ProjectTrackercs10/ProjectTracker.Library/ResourceAssignments.cs
using System;
using System.Data;
using System.Data.SqlClient;
using CSLA;
using CSLA.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
	public class ResourceAssignments : BusinessCollectionBase
	{
    #region Business Properties and Methods

    public ResourceAssignment this [int index]
    {
      get
      {
        return (ResourceAssignment)List[index];
      }
    }

    public ResourceAssignment this [Guid projectID]
    {
      get
      {
        foreach(ResourceAssignment r in List)
          if(r.ProjectID.Equals(projectID))
            return r;
        return null;
      }
    }

    public void AssignTo(Project project, string role)
    {
      DoAssignment(ResourceAssignment.NewResourceAssignment(project, role));
    }

    public void AssignTo(Guid projectID, string role)
    {
      DoAssignment(ResourceAssignment.NewResourceAssignment(projectID, role));
    }

    public void AssignTo(Guid projectID)
    {
      DoAssignment(ResourceAssignment.NewResourceAssignment(projectID));
    }

    private void DoAssignment(ResourceAssignment project)
    {
      if(!Contains(project))
        List.Add(project);
      else
        throw new Exception("Resource already assigned to project");
    }

    public void Remove(ResourceAssignment project)
    {
      List.Remove(project);
    }

    public void Remove(Guid projectID)
    {
      foreach(ResourceAssignment obj in List)
        if(obj.ProjectID.Equals(projectID))
        {
          Remove(obj);
          break;
        }
    }

    #endregion

    #region Contains

    public bool Contains(ResourceAssignment assignment)
    {
      foreach(ResourceAssignment child in List)
        if(child.Equals(assignment))
          return true;
      return false;
    }

    public bool ContainsDeleted(ResourceAssignment assignment)
    {
      foreach(ResourceAssignment child in deletedList)
        if(child.Equals(assignment))
          return true;
      return false;
    }

    public bool Contains(Guid projectID)
    {
      foreach(ResourceAssignment r in List)
        if(r.ProjectID.Equals(projectID))
          return true;
      return false;
    }

    public bool ContainsDeleted(Guid projectID)
    {
      foreach(ResourceAssignment r in deletedList)
        if(r.ProjectID.Equals(projectID))
          return true;
      return false;
    }

    #endregion

    #region Shared Methods

    internal static ResourceAssignments NewResourceAssignments()
    {
      return new ResourceAssignments();
    }

    internal static ResourceAssignments GetResourceAssignments(SafeDataReader dr)
    {
      ResourceAssignments col = new ResourceAssignments();
      col.Fetch(dr);
      return col;
    }

    #endregion

    #region Constructors

		public ResourceAssignments()
		{
      MarkAsChild();
		}

    #endregion

    #region Data Access

    // called by Resource to load data from the database
    private void Fetch(SafeDataReader dr)
    {
      while(dr.Read())
        List.Add(ResourceAssignment.GetResourceAssignment(dr));
    }

    // called by Resource to delete/add/update data into the database
    internal void Update(SqlTransaction tr, Resource resource)
    {
      // update (thus deleting) any deleted child objects
      foreach(ResourceAssignment obj in deletedList)
        obj.Update(tr, resource);
      // now that they are deleted, remove them from memory too
      deletedList.Clear();

      // add/update any current child objects
      foreach(ResourceAssignment obj in List)
        obj.Update(tr, resource);
    }

    #endregion

  }
}

File: ProjectTracker10/ProjectTracker.Library/ResourceAssignments.vb
Imports System.Data.SqlClient
Imports CSLA.Data

<Serializable()> _
Public Class ResourceAssignments
  Inherits BusinessCollectionBase

#Region " Business Properties and Methods "

  Default Public ReadOnly Property Item(ByVal Index As Integer) As ResourceAssignment
    Get
      Return CType(list.Item(Index), ResourceAssignment)
    End Get
  End Property

  Default Public ReadOnly Property Item(ByVal ProjectID As Guid) As ResourceAssignment
    Get
      Dim r As ResourceAssignment

      For Each r In list
        If r.ProjectID.Equals(ProjectID) Then
          Return r
        End If
      Next
    End Get
  End Property

  Public Sub AssignTo(ByVal Project As Project, ByVal Role As String)
    DoAssignment(ResourceAssignment.NewResourceAssignment(Project, Role))
  End Sub

  Public Sub AssignTo(ByVal ProjectID As Guid, ByVal Role As String)
    DoAssignment(ResourceAssignment.NewResourceAssignment(ProjectID, Role))
  End Sub

  Public Sub AssignTo(ByVal ProjectID As Guid)
    DoAssignment(ResourceAssignment.NewResourceAssignment(ProjectID))
  End Sub

  Private Sub DoAssignment(ByVal Project As ResourceAssignment)
    If Not Contains(Project) Then

      list.Add(Project)

    Else
      Throw New Exception("Resource already assigned to project")
    End If
  End Sub

  Public Sub Remove(ByVal Project As ResourceAssignment)
    list.Remove(Project)
  End Sub

  Public Sub Remove(ByVal ProjectID As Guid)
    Dim obj As ResourceAssignment

    For Each obj In list
      If obj.ProjectID.Equals(ProjectID) Then
        Remove(obj)
        Exit For
      End If
    Next
  End Sub

#End Region

#Region " Contains "

  Public Overloads Function Contains( _
      ByVal Assignment As ResourceAssignment) As Boolean

    Dim child As ResourceAssignment

    For Each child In list
      If child.Equals(Assignment) Then
        Return True
      End If
    Next

    Return False

  End Function

  Public Overloads Function ContainsDeleted( _
      ByVal Assignment As ResourceAssignment) As Boolean

    Dim child As ResourceAssignment

    For Each child In deletedList
      If child.Equals(Assignment) Then
        Return True
      End If
    Next

    Return False

  End Function

  Public Overloads Function Contains( _
      ByVal ProjectID As Guid) As Boolean

    Dim r As ResourceAssignment

    For Each r In list
      If r.ProjectID.Equals(ProjectID) Then
        Return True
      End If
    Next

    Return False

  End Function

  Public Overloads Function ContainsDeleted( _
    ByVal ProjectID As Guid) As Boolean

    Dim r As ResourceAssignment

    For Each r In deletedList
      If r.ProjectID.Equals(ProjectID) Then
        Return True
      End If
    Next

    Return False

  End Function

#End Region

#Region " Shared Methods"

  Friend Shared Function NewResourceAssignments() As ResourceAssignments
    Return New ResourceAssignments
  End Function

  Friend Shared Function GetResourceAssignments( _
          ByVal dr As SafeDataReader) As ResourceAssignments
    Dim col As New ResourceAssignments
    col.Fetch(dr)
    Return col
  End Function

#End Region

#Region " Constructors "

  Private Sub New()
    ' allow creation only from parent
    MarkAsChild()
  End Sub

#End Region

#Region " Data Access "

  ' called by Resource to load data from the database
  Private Sub Fetch(ByVal dr As SafeDataReader)
    While dr.Read()
      list.Add(ResourceAssignment.GetResourceAssignment(dr))
    End While
  End Sub

  ' called by Resource to delete/add/update data into the database
  Friend Sub Update(ByVal tr As SqlTransaction, ByVal Resource As Resource)
    Dim obj As ResourceAssignment

    ' update (thus deleting) any deleted child objects
    For Each obj In deletedList
      obj.Update(tr, Resource)
    Next
    ' now that they are deleted, remove them from memory too
    deletedList.Clear()

    ' add/update any current child objects
    For Each obj In list
      obj.Update(tr, Resource)
    Next

  End Sub

#End Region

End Class
